Update raw data

josenavas opened the pull request:
Adds a command to be able to update the files attached to a raw data object - similar to what we have for update preprocessed data.

This is needed to update the EMP data uploaded at the beginning of Qiita, and this can be useful for any future problems.

@antgonza able to review?

antgonza commented:
My only concern is that update_raw_data_from_cmd overwrites (deletes) old raw data, should we have a parameter or confirmation for this? Note that I understand why this behavior is required for the current situation but not sure we want to have it in the future.

ElDeveloper commented on qiita_db/commands.py:
```diff
@@ -307,6 +307,65 @@ def load_parameters_from_cmd(name, fp, table):
+        If rd_id is provided and it does not belong to the given study
+    """
+    if len(filepaths) != len(filepath_types):
+        raise ValueError("Please provide exactly one filepath_type for each"
```
There's a space missing at the end of this sentence.

josenavas replied:
Good catch.

ElDeveloper commented on qiita_db/commands.py:
```diff
@@ -307,6 +307,65 @@ def load_parameters_from_cmd(name, fp, table):
+        study = Study(study_id)
+        raw_data_ids = study.raw_data()
+        if not raw_data_ids:
+            raise ValueError("Study %s does not have any raw data" % study_id)
```
study_id is an int right so it should really be a %d ... though I know python will just coerce it to a string.

josenavas replied:
Done

ElDeveloper commented on qiita_db/commands.py:
```diff
@@ -307,6 +307,65 @@ def load_parameters_from_cmd(name, fp, table):
+            raw_data = RawData(sorted(raw_data_ids)[0])
+
+        filepath_types_dict = get_filepath_types()
+        filepath_types = [filepath_types_dict[x] for x in filepath_types]
```
In the extreme case that a filepath type didn't exist, this would raise a `KeyError` right?

josenavas replied:
Uhm, you're correct. The main usage of this function is through the CLI and in there, those values can only be the correct ones (since the option is of type Choice).

Do you want me to put a specific try/except?

ElDeveloper replied:
I don't have a strong feeling on this, what I worry is when this gets
used elsewhere and something goes wrong, the only error we are going to
get is a `KeyError: 'foo'`.

On (Sep-30-15|15:32), Jose Navas wrote:
> > +        raw_data_ids = study.raw_data()
> > +        if not raw_data_ids:
> > +            raise ValueError("Study %s does not have any raw data" % study_id)
> > +
> > +        if rd_id:
> > +            if rd_id not in raw_data_ids:
> > +                raise ValueError(
> > +                    "The raw data %d does not exist in the study %d. Available"
> > +                    " raw data: %s"
> > +                    % (rd_id, study_id, ', '.join(map(str, raw_data_ids))))
> > +            raw_data = RawData(rd_id)
> > +        else:
> > +            raw_data = RawData(sorted(raw_data_ids)[0])
> > +
> > +        filepath_types_dict = get_filepath_types()
> > +        filepath_types = [filepath_types_dict[x] for x in filepath_types]
> 
> Uhm, you're correct. The main usage of this function is through the CLI and in there, those values can only be the correct ones (since the option is of type Choice).
> 
> Do you want me to put a specific try/except?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/qiita/pull/1482/files#r40861090

josenavas replied:
Done

ElDeveloper commented on scripts/qiita:
```diff
@@ -279,6 +280,25 @@ def load_parameters(fp, table, name):
+@click.option('--fp_type', required=True, multiple=True, help='Describes the '
+              'contents of the file. Pass one fp_type per fp.',
+              type=click.Choice(get_filepath_types().keys()))
+@click.option('--study', required=True, type=int,
```
Is it safe to assume the type is going to be > 0? if that's the case range options may be useful http://click.pocoo.org/5/options/#range-options

josenavas replied:
Good call - done

ElDeveloper commented on scripts/qiita:
```diff
@@ -279,6 +280,25 @@ def load_parameters(fp, table, name):
+              type=click.Choice(get_filepath_types().keys()))
+@click.option('--study', required=True, type=int,
+              help='Study whose raw data will be updated')
+@click.option('--raw_data', required=False, type=int,
```
Same as above.

josenavas replied:
Done

ElDeveloper commented:
Is it worthwhile to add a note the change log?

josenavas commented:
@ElDeveloper I've addressed all your comments.
@antgonza That is not the route that we used for the `update_preprocessed_data`. Note that here it is not removing the data, is actually moving it to the uploads folder, so we are not loosing any data.

This is part of the CLI which is supposed to be used by advanced users/

ElDeveloper commented on qiita_db/commands.py:
```diff
@@ -307,6 +307,73 @@ def load_parameters_from_cmd(name, fp, table):
+
+    Returns
+    -------
+    qiita_db.data.RawData
```
Can you add a description for this return value?

ElDeveloper commented:
Thanks @josenavas, it looks good to me now!

On (Sep-30-15|16:02), Jose Navas wrote:
> @ElDeveloper I've addressed all your comments.
> @antgonza That is not the route that we used for the `update_preprocessed_data`. Note that here it is not removing the data, is actually moving it to the uploads folder, so we are not loosing any data.
> 
> This is part of the CLI which is supposed to be used by advanced users/
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/qiita/pull/1482#issuecomment-144569838

antgonza commented:
:+1:
